File: src/main/java/caso3/lp3/riemannSum.js
class RiemannSum {
    constructor(polynomial, a, b, n) {
        this.terms = this.parsePolynomial(polynomial);
        this.coefficients = this.getCoefficients(this.terms);
        this.exponents = this.getExponents(this.terms);
        this.a = a;
        this.b = b;
        this.n = n;
    }

    parsePolynomial(polynomial) {
        const termRegex = /(?<!\^)(?=[+-](?!-))/;
        let terms = polynomial.split(termRegex);
        
        for (let i = 0; i < terms.length; i++) {
            terms[i] = terms[i].trim();
            
            if (terms[i].endsWith("^")) {
                terms[i] += terms[i + 1];
                terms[i + 1] = "";
            }
        }
        
        return terms.filter(term => term !== "");
    }

    getCoefficients(terms) {
        const coefficients = new Array(terms.length);
        const coefficientPattern = /([-+]?\d+)/;
        
        for (let i = 0; i < terms.length; i++) {
            const term = terms[i];
            const coefficientMatcher = term.match(coefficientPattern);
        
            if (coefficientMatcher) {
                let coefficient;
                if (term.includes("x")) {
                    coefficient = coefficientMatcher[1] || (term.startsWith("-") ? "-1" : "1");
                } else {
                    coefficient = coefficientMatcher[1];
                }
                coefficients[i] = parseFloat(coefficient);
                } else {
                    coefficients[i] = 0.0;
            }
        }
        
        return coefficients;
    }

    getExponents(terms) {
        const exponents = new Array(terms.length);
        const exponentPattern = /x\^([-+]?\d+)/;
        
        for (let i = 0; i < terms.length; i++) {
            const term = terms[i];
            const exponentMatcher = term.match(exponentPattern);
            
            if (exponentMatcher) {
                exponents[i] = parseInt(exponentMatcher[1]);
            } else {
                exponents[i] = 0;
            }
        }
        
        return exponents;
    }
}

class RiemannSumThread {
    constructor(terms, coefficients, exponents, a, b, numIntervals, thread_index) {
        this.terms = terms;
        this.coefficients = coefficients;
        this.exponents = exponents;
        this.a = a;
        this.b = b;
        this.numIntervals = numIntervals;
        this.thread_index = thread_index;
    }

    start() {
        this.calculatePartialSum();
        console.log(`Hilo (${this.thread_index + 1}) | a=${this.a} - b=${this.b} - n=${this.numIntervals} - area=${this.sum}`);
    }

    join() {
        return new Promise((resolve, reject) => {
            const interval = setInterval(() => {
                if (this.isPartialSumReady()) {
                    clearInterval(interval);
                    resolve();
                }
            }, 100);
        });
    }

    calculatePartialSum() {
        let sum = 0.0;
        const delta_x = (this.b - this.a) / this.numIntervals;

        for (let i = 0; i < this.numIntervals; i++) {
            const x = this.a + (i + 0.5) * delta_x;
            const y = this.evaluatePolynomial(x);
            sum += y * delta_x;
        }
        this.sum = sum;
    }

    evaluatePolynomial(x) {
        let result = 0.0;
        for (let i = 0; i < this.terms.length; i++) {
            const coefficient = this.coefficients[i];
            const exponent = this.exponents[i];
            result += coefficient * Math.pow(x, exponent);
        }
        return result;
    }


    isPartialSumReady() {
        return typeof this.sum !== 'undefined';
    }
}

module.exports = {
    RiemannSum,
    RiemannSumThread,
};

File: src/main/java/caso2/lp2/riemann_sum.py
import re
import threading


class RiemannSum:
    def __init__(self, a, b, n, polinomio):
        self.a = a
        self.b = b
        self.n = n
        self.terms = self.get_terms(polinomio)
        self.coefficients = self.get_coefficients(self.terms)
        self.exponents = self.get_exponents(self.terms)

    @staticmethod
    def get_terms(polynomial):
        term_regex = r"(?<!\^)(?=[+-](?!-))"
        terms = re.split(term_regex, polynomial)
        
        for i in range(len(terms)):
            terms[i] = terms[i].strip()
            
            if terms[i].endswith("^"):
                terms[i] += terms[i + 1]
                terms[i + 1] = ""
        
        return list(filter(lambda term: term != "", terms))

    @staticmethod
    def get_coefficients(terms):
        coefficients = [None] * len(terms)
        coefficient_pattern = r"([-+]?\d+)"
        
        for i in range(len(terms)):
            term = terms[i]
            coefficient_matcher = re.search(coefficient_pattern, term)
            
            if coefficient_matcher:
                coefficient = coefficient_matcher.group(1)
                
                if "x" in term:
                    coefficient = coefficient or ("-1" if term.startswith("-") else "1")
                
                coefficients[i] = float(coefficient)
            else:
                coefficients[i] = 0.0
        
        return coefficients

    @staticmethod
    def get_exponents(terms):
        exponents = [None] * len(terms)
        exponent_pattern = r"x\^([-+]?\d+)"
        
        for i in range(len(terms)):
            term = terms[i]
            exponent_matcher = re.search(exponent_pattern, term)
            
            if exponent_matcher:
                exponents[i] = int(exponent_matcher.group(1))
            else:
                exponents[i] = 0
        
        return exponents


class RiemannSumThread(threading.Thread):
    def __init__(self, terms, coefficients, exponents, a, b, num_intervals, thread_index, sums):
        threading.Thread.__init__(self)
        self.terms = terms
        self.coefficients = coefficients
        self.exponents = exponents
        self.a = a
        self.b = b
        self.num_intervals = num_intervals
        self.thread_index = thread_index
        self.sums = sums

    def run(self):
        sum_result = self.calculate_sum()
        self.sums[self.thread_index] = sum_result

    def calculate_sum(self):
        delta_x = float(self.b - self.a) / self.num_intervals
        sum_result = 0.0

        for i in range(self.num_intervals):
            x = self.a + (i + 0.5) * delta_x
            term_sum = 0.0

            for j in range(len(self.terms)):
                term_sum += self.coefficients[j] * (x ** self.exponents[j])

            sum_result += term_sum * delta_x

        return sum_result
